### PR TITLE
Fixes #12441 validate uniqueness for nic identifier and host

### PR DESCRIPTION
**Pull request by brandonweeks**
Taking a stab at fixing #12441. This helps prevent an issue where a when the mac address of a nic changes a duplicate row is inserted with the same identifier, which causes validation problems in the web interface and doesn't make a terrible amount of sense.

### CODE REVIEW

**Comment by brandonweeks** [comment]
This PR successfully prevents the duplicate nic from being inserted, alas we then get this:

> Saving br_tun NIC for host $host failed, skipping because:
> Identifier has already been taken

Ideally the the existing interface would be updated with the new MAC address.

Edit: This seems to me like a different issue entirely, so I've filed it as #12443.

**Comment by domcleal** [comment]
Thanks for filing the PR, running tests now: [test]

**Comment by mmoll** [comment]
[test]

**Comment by brandonweeks** [comment]
Updated the commit to pass all tests. The validation is a little more narrowly scoped now but still prevents the specific issue I documented in #12441.

File: test/unit/nics/base_test.rb [hunk 1]
```diff
@@ -98,5 +98,18 @@ class NicBaseTest < ActiveSupport::TestCase
...
+      let(:nic) do
+        nic = host.interfaces.build(:managed => true, :type => 'Nic::Managed')
+        nic.identifier = host.primary_interface.identifier
+        nic
+      end
+
+      test 'it is invalid because of conflicting identifier' do
+        refute nic.valid?
+        assert nic.errors.has_key?(:identifier)
+      end
```
Comment (domcleal) [inline test/unit/nics/base_test.rb]: The rubocop lint check is failing on this line due to trailing whitespace: http://ci.theforeman.org/job/test_develop_pr_rubocop/4816/checkstyleResult/new/

**Comment by mmoll** [comment]
[test]

**Comment by brandonweeks** [comment]
One last time. Improved the validation a little by skipping validation only in Host::Base.set_interfaces instead of on all primary interfaces.

**Comment by mmoll** [comment]
[test]

**Comment by domcleal** [comment]
@brandonweeks thanks for accepting the organisation invite.  If you make your membership public on [this page](https://github.com/orgs/theforeman/people) then next time you update the PR, Jenkins will automatically test it.

File: app/models/nic/base.rb [hunk 2]
```diff
@@ -27,6 +27,10 @@ class Base < ActiveRecord::Base
 
     validates :host, :presence => true, :if => Proc.new { |nic| nic.require_host? }
 
+    validates :identifier, :uniqueness => { :scope => :host_id },
+              :if => Proc.new { |nic| nic.identifier && nic.host },
```
Comment (dLobatog) [inline app/models/nic/base.rb:31]: I'll change these procs to stabby lambda on merge

**Comment by dLobatog** [comment]
Merged as fedf37919ea9df65b40d39dc0387365d2c9e703d, thanks @brandonweeks

File: app/models/host/base.rb [hunk 3]
```diff
@@ -164,7 +164,7 @@ def set_interfaces(parser)
       if !self.managed? && self.primary_interface.mac.blank? && self.primary_interface.identifier.blank?
         identifier, values = parser.suggested_primary_interface(self)
         self.primary_interface.mac = Net::Validations.normalize_mac(values[:macaddress]) if values.present?
-        self.primary_interface.identifier = identifier
+        self.primary_interface.update_attribute(:identifier, identifier)
```
Comment (ares) [inline app/models/host/base.rb:167]: why was this needed? it's causing double save of primary interface, this even skips the validation

Comment (dLobatog) [inline app/models/host/base.rb:167]: See http://projects.theforeman.org/issues/12441 for a reproducer , it's preventing editing the host after fact upload in some cases
